In [117]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt




In [118]:
df=pd.read_csv('bengaluru_house_prices.csv')

In [119]:
df.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [120]:
df['availability'].unique()

array(['19-Dec', 'Ready To Move', '18-May', '18-Feb', '18-Nov', '20-Dec',
       '17-Oct', '21-Dec', '19-Sep', '20-Sep', '18-Mar', '20-Feb',
       '18-Apr', '20-Aug', '18-Oct', '19-Mar', '17-Sep', '18-Dec',
       '17-Aug', '19-Apr', '18-Jun', '22-Dec', '22-Jan', '18-Aug',
       '19-Jan', '17-Jul', '18-Jul', '21-Jun', '20-May', '19-Aug',
       '18-Sep', '17-May', '17-Jun', '21-May', '18-Jan', '20-Mar',
       '17-Dec', '16-Mar', '19-Jun', '22-Jun', '19-Jul', '21-Feb',
       'Immediate Possession', '19-May', '17-Nov', '20-Oct', '20-Jun',
       '19-Feb', '21-Oct', '21-Jan', '17-Mar', '17-Apr', '22-May',
       '19-Oct', '21-Jul', '21-Nov', '21-Mar', '16-Dec', '22-Mar',
       '20-Jan', '21-Sep', '21-Aug', '14-Nov', '19-Nov', '15-Nov',
       '16-Jul', '15-Jun', '17-Feb', '20-Nov', '20-Jul', '16-Sep',
       '15-Oct', '15-Dec', '16-Oct', '22-Nov', '15-Aug', '17-Jan',
       '16-Nov', '20-Apr', '16-Jan', '14-Jul'], dtype=object)

In [121]:
df['society'].unique()

array(['Coomee ', 'Theanmp', nan, ..., 'SJovest', 'ThhtsV ', 'RSntsAp'],
      dtype=object)

In [122]:
len(df['society'].unique())

2689

In [123]:
df.shape

(13320, 9)

In [124]:
df2=df.drop(['area_type','availability','society','balcony'],axis=1)

In [125]:
df2.shape

(13320, 5)

### Handling Missing values

In [126]:
df2.isnull().sum()

location       1
size          16
total_sqft     0
bath          73
price          0
dtype: int64

In [127]:
df2.dropna(inplace=True)

In [128]:
df2.isnull().sum()

location      0
size          0
total_sqft    0
bath          0
price         0
dtype: int64

In [129]:
df2.shape

(13246, 5)

## Feature Engineering

In [130]:
df2['size'].unique()

array(['2 BHK', '4 Bedroom', '3 BHK', '4 BHK', '6 Bedroom', '3 Bedroom',
       '1 BHK', '1 RK', '1 Bedroom', '8 Bedroom', '2 Bedroom',
       '7 Bedroom', '5 BHK', '7 BHK', '6 BHK', '5 Bedroom', '11 BHK',
       '9 BHK', '9 Bedroom', '27 BHK', '10 Bedroom', '11 Bedroom',
       '10 BHK', '19 BHK', '16 BHK', '43 Bedroom', '14 BHK', '8 BHK',
       '12 Bedroom', '13 BHK', '18 Bedroom'], dtype=object)

##### add new feature is BHK

In [131]:
df2['bhk']=df2['size'].apply(lambda x:x.split(' ')[0])

In [132]:
df2['bhk'].info()

<class 'pandas.core.series.Series'>
Index: 13246 entries, 0 to 13319
Series name: bhk
Non-Null Count  Dtype 
--------------  ----- 
13246 non-null  object
dtypes: object(1)
memory usage: 207.0+ KB


In [133]:
df2['bhk']=df2['bhk'].astype('int')

In [134]:
df2.head()

location       size total_sqft  bath   price  bhk
0  Electronic City Phase II      2 BHK       1056   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom       2600   5.0  120.00    4
2               Uttarahalli      3 BHK       1440   2.0   62.00    3
3        Lingadheeranahalli      3 BHK       1521   3.0   95.00    3
4                  Kothanur      2 BHK       1200   2.0   51.00    2

In [135]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13246 entries, 0 to 13319
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13246 non-null  object 
 1   size        13246 non-null  object 
 2   total_sqft  13246 non-null  object 
 3   bath        13246 non-null  float64
 4   price       13246 non-null  float64
 5   bhk         13246 non-null  int64  
dtypes: float64(2), int64(1), object(3)
memory usage: 724.4+ KB


## Feature Englineering

In [136]:
df2['total_sqft'].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

In [137]:
#### here we want to change the data types and change the range to a value

In [138]:
def is_float(x):
    try:
        float(x)
    except:
        return False
    return True
    

In [139]:
df2[~df2['total_sqft'].apply(is_float)]

location   size   total_sqft  bath    price  bhk
30              Yelahanka  4 BHK  2100 - 2850   4.0  186.000    4
122                Hebbal  4 BHK  3067 - 8156   4.0  477.000    4
137    8th Phase JP Nagar  2 BHK  1042 - 1105   2.0   54.005    2
165              Sarjapur  2 BHK  1145 - 1340   2.0   43.490    2
188              KR Puram  2 BHK  1015 - 1540   2.0   56.800    2
...                   ...    ...          ...   ...      ...  ...
12975          Whitefield  2 BHK   850 - 1060   2.0   38.190    2
12990      Talaghattapura  3 BHK  1804 - 2273   3.0  122.000    3
13059              Harlur  2 BHK  1200 - 1470   2.0   72.760    2
13265               Hoodi  2 BHK  1133 - 1384   2.0   59.135    2
13299          Whitefield  4 BHK  2830 - 2882   5.0  154.500    4

[190 rows x 6 columns]

##### here we can see in total_sqft column there are many values in a range like 2100 - 2850 , we have to replace this value with its average and we can convert it into float , and also it may have some string values like 3410sqft , we should handle this kind of problem . following function will convert sqft to number.

In [140]:
def convert_sqft_to_num(x):
    tokens = x.split('-')
    if len(tokens) == 2:
        return (float(tokens[0])+float(tokens[1]))/2
    try:
        return float(x)
    except:
        return None   

        

In [141]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13246 entries, 0 to 13319
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13246 non-null  object 
 1   size        13246 non-null  object 
 2   total_sqft  13246 non-null  object 
 3   bath        13246 non-null  float64
 4   price       13246 non-null  float64
 5   bhk         13246 non-null  int64  
dtypes: float64(2), int64(1), object(3)
memory usage: 724.4+ KB


In [142]:
df3 = df2.copy()
df3.total_sqft = df3.total_sqft.apply(convert_sqft_to_num)
df3 = df3[df3.total_sqft.notnull()]
df3.head(2)


location       size  total_sqft  bath   price  bhk
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4

In [143]:
df3.shape

(13200, 6)

In [144]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13200 entries, 0 to 13319
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13200 non-null  object 
 1   size        13200 non-null  object 
 2   total_sqft  13200 non-null  float64
 3   bath        13200 non-null  float64
 4   price       13200 non-null  float64
 5   bhk         13200 non-null  int64  
dtypes: float64(3), int64(1), object(2)
memory usage: 721.9+ KB


## Feature Englineering

In [145]:
df3.head()

location       size  total_sqft  bath   price  bhk
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2

#### create a new column price_per_total_sqft

In [146]:
df4=df3.copy()
df4['price_per_total_sqft']=df4['price']*100000/df4['total_sqft']

In [147]:
df4

location       size  total_sqft  bath   price  bhk  \
0      Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1              Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2                   Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3            Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                      Kothanur      2 BHK      1200.0   2.0   51.00    2   
...                         ...        ...         ...   ...     ...  ...   
13315                Whitefield  5 Bedroom      3453.0   4.0  231.00    5   
13316             Richards Town      4 BHK      3600.0   5.0  400.00    4   
13317     Raja Rajeshwari Nagar      2 BHK      1141.0   2.0   60.00    2   
13318           Padmanabhanagar      4 BHK      4689.0   4.0  488.00    4   
13319              Doddathoguru      1 BHK       550.0   1.0   17.00    1   

       price_per_total_sqft  
0               3699.810606  
1               4615.384615  
2               4305.555556  
3               6245.890861  
4               4250.000000  
...                     ...  
13315           6689.834926  
13316          11111.111111  
13317           5258.545136  
13318          10407.336319  
13319           3090.909091  

[13200 rows x 7 columns]

In [148]:
df3.describe()

total_sqft          bath         price           bhk
count  13200.000000  13200.000000  13200.000000  13200.000000
mean    1555.302783      2.691136    112.276178      2.800833
std     1237.323445      1.338915    149.175995      1.292843
min        1.000000      1.000000      8.000000      1.000000
25%     1100.000000      2.000000     50.000000      2.000000
50%     1275.000000      2.000000     71.850000      3.000000
75%     1672.000000      3.000000    120.000000      3.000000
max    52272.000000     40.000000   3600.000000     43.000000

In [149]:
df4.to_csv('bhp.csv',index=False)

In [150]:
df4.location.value_counts(ascending=False)

location
Whitefield                       532
Sarjapur  Road                   392
Electronic City                  302
Kanakpura Road                   264
Thanisandra                      232
                                ... 
Gubbi Cross, Hennur Main Road      1
Sabari Nagar                       1
Ashraya Layout                     1
Whitefield ECC Road                1
Maruthi nagar kogilu               1
Name: count, Length: 1298, dtype: int64

In [151]:
df4.location = df4.location.apply(lambda x: x.strip())
location_stats = df4['location'].value_counts(ascending=False)
location_stats

location
Whitefield                               533
Sarjapur  Road                           392
Electronic City                          304
Kanakpura Road                           264
Thanisandra                              235
                                        ... 
Kanakapura main  Road                      1
Masjid e Alkareem                          1
Sindhi Colony                              1
Kanakapur main road                        1
Bidere Agarahara, Behind Safal market      1
Name: count, Length: 1287, dtype: int64

In [152]:
df4.shape

(13200, 7)

In [153]:
len(location_stats)

1287

In [154]:
len(location_stats[location_stats>10])

240

In [155]:
len(location_stats[location_stats<=10])

1047

### Dimensionality Reduction

##### here there are a lot of categorical variable in location column what are the number of categorical variable is less than 10 , we might catagorise it as 'other' category.

In [156]:
location_status_lessthan_10=location_stats[location_stats<=10]

In [157]:
len(df4.location.unique())

1287

In [158]:
df4.location=df4.location.apply(lambda x:'other' if x in location_status_lessthan_10 else x)

In [159]:
len(df4.location.unique())

241

In [160]:
df4.head(10)

location       size  total_sqft  bath   price  bhk  \
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2   
5                Whitefield      2 BHK      1170.0   2.0   38.00    2   
6          Old Airport Road      4 BHK      2732.0   4.0  204.00    4   
7              Rajaji Nagar      4 BHK      3300.0   4.0  600.00    4   
8              Marathahalli      3 BHK      1310.0   3.0   63.25    3   
9                     other  6 Bedroom      1020.0   6.0  370.00    6   

   price_per_total_sqft  
0           3699.810606  
1           4615.384615  
2           4305.555556  
3           6245.890861  
4           4250.000000  
5           3247.863248  
6           7467.057101  
7          18181.818182  
8           4828.244275  
9          36274.509804

### Removing OutLiers 

##### Consider the size of a room(1 bhk) is less than 300 sqft as outliers .

In [161]:
df4.shape

(13200, 7)

In [162]:
df4.head()

location       size  total_sqft  bath   price  bhk  \
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2   

   price_per_total_sqft  
0           3699.810606  
1           4615.384615  
2           4305.555556  
3           6245.890861  
4           4250.000000

In [163]:
df4.bhk.unique()

array([ 2,  4,  3,  6,  1,  8,  7,  5, 11,  9, 27, 10, 19, 16, 43, 14, 12,
       13, 18])

In [164]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13200 entries, 0 to 13319
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   location              13200 non-null  object 
 1   size                  13200 non-null  object 
 2   total_sqft            13200 non-null  float64
 3   bath                  13200 non-null  float64
 4   price                 13200 non-null  float64
 5   bhk                   13200 non-null  int64  
 6   price_per_total_sqft  13200 non-null  float64
dtypes: float64(4), int64(1), object(2)
memory usage: 825.0+ KB


In [165]:
df4[~(df4.total_sqft/df4.bhk<300)]

location       size  total_sqft  bath   price  bhk  \
0      Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1              Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2                   Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3            Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                      Kothanur      2 BHK      1200.0   2.0   51.00    2   
...                         ...        ...         ...   ...     ...  ...   
13315                Whitefield  5 Bedroom      3453.0   4.0  231.00    5   
13316                     other      4 BHK      3600.0   5.0  400.00    4   
13317     Raja Rajeshwari Nagar      2 BHK      1141.0   2.0   60.00    2   
13318           Padmanabhanagar      4 BHK      4689.0   4.0  488.00    4   
13319              Doddathoguru      1 BHK       550.0   1.0   17.00    1   

       price_per_total_sqft  
0               3699.810606  
1               4615.384615  
2               4305.555556  
3               6245.890861  
4               4250.000000  
...                     ...  
13315           6689.834926  
13316          11111.111111  
13317           5258.545136  
13318          10407.336319  
13319           3090.909091  

[12456 rows x 7 columns]

In [166]:
## removing outliers 
df4.shape

(13200, 7)

In [167]:
df5=df4[~(df4.total_sqft/df4.bhk<300)]

In [168]:
df5.head()

location       size  total_sqft  bath   price  bhk  \
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2   
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4   
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3   
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3   
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2   

   price_per_total_sqft  
0           3699.810606  
1           4615.384615  
2           4305.555556  
3           6245.890861  
4           4250.000000

In [169]:
df5.shape

(12456, 7)

## Outliers Removing based on mean and deviation

In [170]:
df5.price_per_total_sqft.describe()

count     12456.000000
mean       6308.502826
std        4168.127339
min         267.829813
25%        4210.526316
50%        5294.117647
75%        6916.666667
max      176470.588235
Name: price_per_total_sqft, dtype: float64

#### here we can see that minimum price is 267 and max is 176470 , we want to remove these outliers based on mean and standered deviation in each location.

In [171]:
def remove_outliers_mean_std(df):
    
    df_out=pd.DataFrame()
    for key,subdf in df.groupby('location'):
        mean=np.mean(subdf.price_per_total_sqft)
        std=np.std(subdf.price_per_total_sqft)
        reduced_df=subdf[(subdf.price_per_total_sqft>=(mean-std))&(subdf.price_per_total_sqft<=(mean+std))]
        df_out=pd.concat([df_out,reduced_df],ignore_index=True)
    return df_out

In [172]:
df6=remove_outliers_mean_std(df5)

In [173]:
df6.shape

(10242, 7)

In [174]:
df6.head()

location   size  total_sqft  bath  price  bhk  \
0  1st Block Jayanagar  4 BHK      2850.0   4.0  428.0    4   
1  1st Block Jayanagar  3 BHK      1630.0   3.0  194.0    3   
2  1st Block Jayanagar  3 BHK      1875.0   2.0  235.0    3   
3  1st Block Jayanagar  3 BHK      1200.0   2.0  130.0    3   
4  1st Block Jayanagar  2 BHK      1235.0   2.0  148.0    2   

   price_per_total_sqft  
0          15017.543860  
1          11901.840491  
2          12533.333333  
3          10833.333333  
4          11983.805668

#### we are going to remove columns size and price_per_totak_sqft , because similar values with bhk and price

In [ ]:
df7=df6.drop(['